### PR TITLE
generic ephemeral volume beta

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

We want to promote generic ephemeral volumes to beta in 1.21.

#### Does this PR introduce a user-facing change?
```release-note
Generic ephemeral volumes are beta.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes
- [Usage]: https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes
```

### CODE REVIEW

**Comment by pohly** [comment]
/sig storage
/label api-review
/hold

This PR is ready for testing and review, but should not be merged until several other PRs are also merged - see tracking project in https://github.com/orgs/kubernetes-csi/projects/34

**Comment by pohly** [comment]
/assign @gnufied

Can you review for SIG storage?

**Comment by liggitt** [comment]
btw, I opened https://github.com/kubernetes/kubernetes/issues/99716 to track improving the test fixture generator to avoid triggering these enormous changes to unrelated fields in the HEAD fixtures

File: staging/src/k8s.io/api/core/v1/types.go
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go:1801]: comment out rather than delete, and add a tombstone comment so we never reuse this json field name or proto tag. see other "tombstoned" comments in types.go files for examples.

File: staging/src/k8s.io/api/core/v1/types.go
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go:159]: API doc should call out this is a beta feature and depends on the GenericEphemeralVolume feature gate being enabled

**Review by liggitt — COMMENTED** [review]
have the Alpha → Beta criteria in https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes#alpha---beta-graduation been met?

**Comment by pohly** [comment]
> have the Alpha → Beta criteria in https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes#alpha---beta-graduation been met?

Yes, albeit with some changes still pending in other PRs. Therefore I put this one on hold and will only cancel the hold once they are all in.

File: staging/src/k8s.io/api/core/v1/types.go
Comment (pohly) [reply staging/src/k8s.io/api/core/v1/types.go:1801]: Done.

File: staging/src/k8s.io/api/core/v1/types.go
Comment (pohly) [reply staging/src/k8s.io/api/core/v1/types.go:159]: Done.

**Comment by pohly** [comment]
/assign @saad-ali

For confirmation that:
- removal of `EphemeralVolumeSource.ReadOnly` is indeed what he wanted
- this is meant to become beta

**Comment by liggitt** [comment]
API lgtm, assign once ready for approval

**Comment by pohly** [comment]
I'm trying to get @saad-ali's attention for his okay... :sweat_smile: 

FWIW, the only remaining PR why I still have this one here on hold is https://github.com/kubernetes/kubernetes/pull/98918

**Comment by saad-ali** [comment]
> * removal of EphemeralVolumeSource.ReadOnly is indeed what he wanted

+1 to removal of this ReadOnly field. If we find a volume plugin that doesn't operate correctly without it, we can investigate. Linking https://github.com/kubernetes/kubernetes/issues/70503 for reference

> * this is meant to become beta

Yes the "PVC inline ephemeral volumes (work with CSI driver)" feature is targeted for beta by SIG Storage for the 1.21 release.

/approve
/lgtm

/hold 
To let @liggitt review.

**Comment by pohly** [comment]
/assign @liggitt 

For final approval.

Jordan, do you want to hold this until https://github.com/kubernetes/kubernetes/pull/98918 (the PSP validation PR) is merged or can these two PRs proceed independently?

**Comment by pohly** [comment]
I added one more small change which removed Saad's LGTM from the PR: the latest commit causes the E2E tests to run in the normal prow jobs.

Jordan, if Saad doesn't get to it, can you also add the LGTM to get this merged?

**Comment by pohly** [comment]
I also had to rebase to refresh the `testdata/HEAD/batch.v1.CronJob` files.

**Comment by pohly** [comment]
/hold cancel

All related PRs have merged, this is the last one for this feature.

**Comment by liggitt** [comment]
/approve

for API and client-go changes
storage reviewer has lgtm

**Comment by pohly** [comment]
It looks like the 1.21 milestone still needs to be set for this PR, too.

**Comment by pohly** [comment]
Rebased again with refreshed `staging/src/k8s.io/api/core/v1/generated.pb.go`.

**Review by saad-ali — APPROVED** [review]
/lgtm

**Pull request merged**
